plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/LazyConnectionFactory.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import io.trino.spi.connector.ConnectorSession;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class LazyConnectionFactory
+        implements ConnectionFactory
+{
+    private final ConnectionFactory delegate;
+
+    public LazyConnectionFactory(ConnectionFactory delegate)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        return new LazyConnection(() -> delegate.openConnection(session));
+    }
+
+    @Override
+    public void close()
+            throws SQLException
+    {
+        delegate.close();
+    }
+
+    private static final class LazyConnection
+            extends ForwardingConnection
+    {
+        private final SqlSupplier<Connection> connectionSupplier;
+        private volatile Connection connection;
+
+        public LazyConnection(SqlSupplier<Connection> connectionSupplier)
+        {
+            this.connectionSupplier = requireNonNull(connectionSupplier, "connectionSupplier is null");
+        }
+
+        @Override
+        protected Connection getDelegate()
+                throws SQLException
+        {
+            if (connection == null) {
+                synchronized (this) {
+                    if (connection == null) {
+                        connection = connectionSupplier.get();
+                    }
+                }
+            }
+            return connection;
+        }
+
+        @Override
+        public void close()
+                throws SQLException
+        {
+            if (connection != null) {
+                connection.close();
```

Comment:
If the lazy connection is used after close, it will create a resource leak (unlosed connection).
This is worse than a regular connection use-after-close.

Add `closed` field and consult this in `getDelegate()`. 
Make everything synchronized to keep the code sanely simple

Reply:
> If the lazy connection is used after close,

😱

Reply:
If the worry is that `getDelegate` will re-open the connection, then I don't think that'll ever happen, because the `connection` field is only initialized when it's `null`, and then it's never reset back to `null`. So I think the `closed` field is in fact redundant.

Reply:
good point! Restored oryginal code.

Reply:
consider code:

- open (lazy) connection
- close 
- use it // should fail

it will no longer fail, and it will also leak

Reply:
Ah. There's no "use" stage between "open" and "close". That's the catch! Perhaps `close` could assign some kind of "tombstone" to `delegate`, then.

Reply:
or we just can have a `boolean closed` field for simplicity.

Reply:
Added (restored) field and added a test. I didn't consider that someone might use a connection after `close()`.